fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java

```diff
@@ -265,10 +258,84 @@ public String toString() {
+                                      @Nonnull FieldDescriptorProto.Label protoLabel,
+                                      boolean isNullable) {
+        final var typeCode = TypeCode.fromProtobufType(protoType);
+        if (protoLabel == FieldDescriptorProto.Label.LABEL_REPEATED) {
+            // collection type
+            // case 1: primitive types -- assumed to be non-nullable array of that type
+            if (typeCode.isPrimitive()) {
+                final var primitiveType = primitiveType(typeCode, false);
+                return new Array(primitiveType);
+            } else if (typeCode == TypeCode.ENUM) {
+                final var enumDescriptor = (Descriptors.EnumDescriptor)Objects.requireNonNull(descriptor);
+                final var enumType = new Enum(false, Enum.enumValuesFromProto(enumDescriptor.getValues()));
+                return new Array(enumType);
+            } else {
+                Objects.requireNonNull(descriptor);
+                final var messageDescriptor = (Descriptors.Descriptor)descriptor;
+                // case 2: helper type to model null-ed out array elements
+                final Optional<Descriptors.FieldDescriptor> elementFieldDescriptorMaybe =
+                        Record.arrayElementFieldDescriptorMaybe(messageDescriptor);
+                if (elementFieldDescriptorMaybe.isPresent()) {
+                    final var elementFieldDescriptor = elementFieldDescriptorMaybe.get();
+                    return new Array(fromProtoType(getTypeSpecificDescriptor(elementFieldDescriptor), elementFieldDescriptor.getType(), FieldDescriptorProto.Label.LABEL_OPTIONAL, true));
+                } else {
+                    // case 3: any arbitrary sub message we don't understand
+                    return new Array(fromProtoType(descriptor, protoType, FieldDescriptorProto.Label.LABEL_OPTIONAL, false));
+                }
+            }
+        } else {
+            if (typeCode.isPrimitive()) {
+                return primitiveType(typeCode, isNullable);
+            } else if (typeCode == TypeCode.ENUM) {
+                final var enumDescriptor = (Descriptors.EnumDescriptor)Objects.requireNonNull(descriptor);
+                return new Enum(isNullable, Enum.enumValuesFromProto(enumDescriptor.getValues()));
+            } else if (typeCode == TypeCode.RECORD) {
+                Objects.requireNonNull(descriptor);
+                final var messageDescriptor = (Descriptors.Descriptor)descriptor;
+                return Record.fromFieldDescriptorsMap(isNullable, Record.toFieldDescriptorMap(messageDescriptor.getFields()));
+            }
+        }
+
+        throw new IllegalStateException("unable to translate protobuf descriptor to type");
+    }
+
+    /**
+     * For a given {@link com.google.protobuf.Descriptors.FieldDescriptor} descriptor, returns the message type descriptor
+     * if the field is a message, otherwise <code>null</code>.
```

Comment:
```suggestion
     * For a given {@link com.google.protobuf.Descriptors.FieldDescriptor} descriptor, returns the type descriptor
     * if the field is a message or enum, otherwise <code>null</code>.
```